### PR TITLE
[REF] Remove for-loop over samples in MC Fisher

### LINKED ISSUE #83: Improve efficiency of MC Fisher
Currently, we loop over the data points and MC samples, which is the 'best' way in the sense of FLOPs (`batch_size * mc_samples` backpropagations), but suffers from poor parallelization. We could use `functorch`'s `vmap`, but currently the library relies mostly on `autograd.grad` and I would like to keep it this way as there are certain benefits (e.g. recycling the computation graph when multiplying onto multiple vectors, block-diagonal approximations).

Another way I propose to implement is phrasing multiplication with the MC Fisher as a GGN-vector product where the loss function is such that it's Hessian corresponds to the summed outer product of sampled gradients.
This uses more FLOPs (`batch_size * C` backpropagations), but does not require a loop over `batch_size`. Therefore, this should often yield better run time. One downside of this approach is that it costs as much as multiplying with the exact GGN. This renders the motivation for having an MC-sampled version (less accurate, same cost) weaker.

### PR DESCRIPTION
**Pull request by f-dangel**
Applies the refactoring described in and resolves #83.

Replaces the expensive `autograd` loops over MC samples and data points via a cheaper internal loop over them (which might be refactored further in the future). Computes a GGNVP with a pseudo-loss whose GGN matches the MC Fisher. This approach costs as much as multiplying with the exact Fisher/GGN, but is more parallel and therefore usually much faster.

@runame please check if this fixes your performance problems before reviewing/merging.

### CODE REVIEW

**Review by runame — APPROVED** [review]
Thanks!
